raiden/transfer/channel.py

```diff
@@ -1924,91 +2019,200 @@ def handle_action_set_reveal_timeout(
+        expiration=withdraw_request.expiration,
+        recipient_metadata=withdraw_request.sender_metadata,
+    )
+    channel_state.partner_state.withdraws_pending[withdraw_state.total_withdraw] = withdraw_state
+    channel_state.partner_state.nonce = withdraw_request.nonce
+
+    our_initiated_coop_settle = channel_state.our_state.initiated_coop_settle
+
+    if our_initiated_coop_settle is not None or withdraw_request.coop_settle is True:
+        partner_max_total_withdraw = get_max_withdraw_amount(
+            channel_state.partner_state, channel_state.our_state
+        )
+        if partner_max_total_withdraw != withdraw_request.total_withdraw:
+            return transition_result_invalid(
+                SuccessOrError(
+                    "Partner requested withdraw while we initiated a coop-settle: "
+                    "Partner did not withdraw with maximum balance."
+                )
+            )
+        if get_number_of_pending_transfers(channel_state.partner_state) > 0:
+            return transition_result_invalid(SuccessOrError("Partner has pending transfers."))
+
+        if our_initiated_coop_settle is not None:
+            # There is a coop-settle inplace that we initiated
+            # and partner is communicating their total withdraw with us
+            if not our_initiated_coop_settle.expiration == withdraw_request.expiration:
+                return transition_result_invalid(
+                    SuccessOrError(
+                        "Partner requested withdraw while we initiated a coop-settle: "
+                        "Partner's withdraw has differing expiration."
+                    )
+                )
+
+            msg = "The expected total withdraw of the partner doesn't match the withdraw-request"
+            assert (
+                our_initiated_coop_settle.total_withdraw_participant
+                == withdraw_request.total_withdraw
+            ), msg
+
+            our_initiated_coop_settle.partner_signature_request = withdraw_request.signature
+            coop_settle_events = events_for_coop_settle(
+                channel_state, our_initiated_coop_settle, block_number, block_hash
+            )
+            events.extend(coop_settle_events)
+        else:
+            # Partner initiated a coop-settle
+            our_max_total_withdraw = get_max_withdraw_amount(
+                channel_state.our_state, channel_state.partner_state
+            )
+            if get_number_of_pending_transfers(channel_state.our_state) > 0:
+                return transition_result_invalid(
+                    SuccessOrError("Partner initiated coop-settle, but we have pending transfers.")
+                )
+            partner_initiated_coop_settle = CoopSettleState(
+                total_withdraw_participant=withdraw_request.total_withdraw,
+                total_withdraw_partner=our_max_total_withdraw,
+                expiration=withdraw_request.expiration,
+            )
```

Comment:
Should we have a sanity check for `withdraw_request.expiration`? What happens if the partner sends us a request with unreasonably long expiration?

Reply:
There seems to be little incentive for the peer to do that.
Choosing a very long expiration locks funds for this period. If the peer is the one requesting the coop settle it stands to reason that they are the one with balance to gain.

Also in the worst case either party can always claim on-chain.

Some previous discussions: 
- https://github.com/raiden-network/raiden/issues/864
- https://github.com/raiden-network/raiden/pull/4289